### PR TITLE
AVM Boxes Ops in Pyteal

### PR DESCRIPTION
**Pull request by ahangsu**
Combined with Ben's previous PR and 3 additional ops from JJ's new PR, corresponding ticket https://github.com/algorand/pyteal/issues/300

## Test

`box_test.py` tests the generated Teal code match with expected result, and cannot compile under lower version.

### CODE REVIEW

File: pyteal/ir/ops.py [hunk 1]
```diff
@@ -179,6 +179,13 @@ def min_version(self) -> int:
     gitxnas             = OpType("gitxnas",             Mode.Application,                  6)
     gloadss             = OpType("gloadss",             Mode.Application,                  6)
     acct_params_get     = OpType("acct_params_get",     Mode.Application,                  6)
+    box_create          = OpType("box_create",          Mode.Application,                  7)
```
Comment (michaeldiamant) [inline pyteal/ir/ops.py:182]: @ahangsu A couple of remarks here - Let me know what you think?
* The current expectation is that we'll include Boxes in AVM 8 rather than AVM 7.  So, I think the box opcode versions + `MAX_TEAL_VERSION` should be changed.
* Like for AVM 7, I imagine we'll want a long-lived integration branch for AVM 8.  Can you base the PR off of https://github.com/algorand/pyteal/pull/439?
  * I created https://github.com/algorand/pyteal/pull/439 (`avm8`) from https://github.com/algorand/pyteal/pull/416 (`teal7`).  I'm hoping the proposal minimizes merge conflicts.
  * I'm imagining we'll regularly merge `master` into `teal7` and `avm8`.

File: pyteal/ast/box.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        """
+
+        super().__init__()
+        require_type(name, TealType.bytes)
+        require_type(size, TealType.uint64)
+        self.name = name
+        self.size = size
+
+    def __teal__(self, options: "CompileOptions"):
+        if options.version < Op.box_create.min_version:
```
Comment (michaeldiamant) [inline pyteal/ast/box.py:30]: @ahangsu Few minor thoughts here:
* Via https://github.com/algorand/pyteal/issues/392, we'll soon consolidate TEAL and AVM versioning.  While I'm not asking to pick up the ticket now, I prefer to avoid adding more _TEAL version_ references.
  * As the ticket suggests, depending on the context, _program version_ or _AVM version_ might be better fits. 
* While looking here, it occurred to me a common, helper function exists:  `errors.py::verifyTealVersion`.  Can it be reused?
  * Optional asides - Feel welcomed to disregard:
    * It seems to me the usage of `msg` by callers often duplicates the internal error message.  If you're feeling ambitious, we can discuss normalization options (I see several).
    * When invoking `verifyTealVersion`, I'd prefer keyword arguments for `minVersion` and `version`.  Since the types are the same, I think it's easy to mistakenly invert argument sequencing.  
      * Obviously, a test _should_ exist.  I'm suggesting an optional, secondary safety measure.

File: pyteal/ast/box.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+    def has_return(self):
+        return False
+
+
+BoxCreate.__module__ = "pyteal"
+
+
+class BoxDelete(Expr):
+    """Deletes a box given it's name"""
```
Comment (michaeldiamant) [inline pyteal/ast/box.py:53]: @ahangsu Minor typo - Leaving _suggestion_ with what I think is intended.

```suggestion
    """Deletes a box given its name"""
```

Comment (ahangsu) [reply pyteal/ast/box.py]: good find

File: pyteal/ir/ops.py (hunk 1 above)
Comment (ahangsu) [reply pyteal/ir/ops.py]: just up the version of `MAX_TEAL_VERSION` to 8, and rebased the whole branch against `avm8`

File: pyteal/ast/box.py (hunk 2 above)
Comment (ahangsu) [reply pyteal/ast/box.py]: good point on reusing `verifyTealVersion`, I swapped all implementation to follow such pattern.

On the argument issues for `verifyTealVersion`, if we use keyword arguments for `minVersion` and `version`, then we would have to change a few internal code, and I would like that happening in another separate PR into master.

I am unaware of the consolidating of Teal and AVM version, if AVM version is preferred, then I will try to reflect it in error msg and variable naming.

**Review by barnjamin — APPROVED** [review]

File: pyteal/ast/box.py (hunk 2 above)
Comment (michaeldiamant) [reply pyteal/ast/box.py]: @ahangsu I'm good to resolve.  Thanks for the `verifyTealVersion` change + AVM version consolidation consideration.  Leaving a clarification for informational purposes.

> if we use keyword arguments for minVersion and version, then we would have to change a few internal code

Sorry - I think my vocabulary might be inaccurate.  I'm _not_ considering a change to `verifyTealVersion` implementation.  I'm trying to suggest invoking with named parameters like:

```
verifyTealVersion(
            minVersion = Op.box_del.min_version,
            version = options.version,
            f"{Op.box_del} unavailable for low version.",
)
```

File: pyteal/__init__.pyi [hunk 4]
```diff
@@ -45,6 +45,13 @@ __all__ = [
     "BitwiseOr",
     "BitwiseXor",
     "Block",
+    "BoxCreate",
```
Comment (michaeldiamant) [inline pyteal/__init__.pyi:48]: @ahangsu The functionality looks good to me though I think the user guide ought to reflect boxes in https://pyteal.readthedocs.io/en/stable/state.html.  What's your preference?  Merge as is + open new PR focused only on user guide or include user guide here.

* Either works for me - Looking to gauge the review.
* If helpful to discuss user guide content, I'm happy to talk live - let me know.

Comment (ahangsu) [reply pyteal/__init__.pyi:48]: I would like to finish the last piece of docs in this PR, to let my implementation sit with the document. We should talk about it at some time, that will help me to phrase the documentation better.

File: pyteal/ast/box.py [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+        """
+        super().__init__()
+        require_type(name, TealType.bytes)
+        self.name = name
+
+    def __teal__(self, options: "CompileOptions"):
+        verifyTealVersion(
+            Op.box_del.min_version,
+            options.version,
+            f"{Op.box_del} unavailable for low version.",
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/box.py:68]: oh I see. Then the suggestion makes sense.

File: pyteal/ast/box.py [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+        require_type(size, TealType.uint64)
+        self.name = name
+        self.size = size
+
+    def __teal__(self, options: "CompileOptions"):
+        verifyTealVersion(
+            Op.box_create.min_version,
+            options.version,
+            f"{Op.box_create} unavailable.",
+        )
```
Comment (ahangsu) [inline pyteal/ast/box.py:34]: @michaeldiamant I noticed that, Jacob is working on consolidating TEAL to AVM version, and I am concerned that, the work of making `verifyTealVersion` (`verifyAVMVersion`) arguments to be keywords, would be interrupted because potential merge conflict.

I would like to wait until Jacob's work merged in, and stuffs are updated in avm8 branch according, then start a new PR for this single purpose, for `verifyTealVersion` appears across the codebase.

Comment (ahangsu) [reply pyteal/ast/box.py]: no worries, no change to implementation.

File: docs/state.rst [hunk 7]
```diff
@@ -244,3 +246,119 @@ For example:
...
         If(otherAppOtherAccountRole.hasValue(), otherAppOtherAccountRole.value(), Bytes("none"))
     ])
+
+Box Storage
+-----------
+
+Box storage consists of key-value pairs that are stored in application's local context.
+
+The app account's minimum balance requirement (MBR) is increased by each box, and by the byte size of the box.
+If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account.
```
Comment (tzaffi) [inline docs/state.rst:256]:
```suggestion
WARNING: If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account. It is recommended that _before_ app deletion, all box storage be deleted, and funds previously allocated to the MBR be withdrawn.
```

File: docs/state.rst [hunk 8]
```diff
@@ -244,3 +246,119 @@ For example:
...
+
+Box sizes and names cannot be changed after initial allocation, but they can be deleted and re-allocated.
+Boxes are only visible to the app itself, other apps cannot view the app's own boxes on-chain.
```
Comment (tzaffi) [inline docs/state.rst:259]:
```suggestion
Boxes are only visible to the app itself; in other words, an app cannot read from or write to another app's boxes on-chain.
```

File: docs/state.rst [hunk 9]
```diff
@@ -244,3 +246,119 @@ For example:
...
+For :any:`App.box_create`, the first argument is the box name, and the second argument is the byte size to be allocated.
+
+For example:
+
+.. code-block:: python
+
+    # Allocate a box called "BoxForAlice" of byte size 100
+    App.box_create(Bytes("BoxForAlice"), Int(0xA11CE))
+    # Allocate a box called "BoxForBob" of byte size 90
+    App.box_create(Bytes("BoxForBob"), Int(0xB0B)
```
Comment (tzaffi) [inline docs/state.rst:277]: Similar here, was the intention to just use the commented sizes?

Comment (tzaffi) [reply docs/state.rst:277]: (and a missing right paren)

File: docs/state.rst [hunk 10]
```diff
@@ -244,3 +246,119 @@ For example:
...
+
+For :any:`App.box_put`, the first argument is the box name to create or to write to, and the second argument is the bytes to write.
+
+**Notice**: if the box exists, then :any:`App.box_put` will write the contents to the box;
```
Comment (tzaffi) [inline docs/state.rst:281]:
```suggestion
**Notice**: if the box exists, then :any:`App.box_put` will write the contents to the box (assuming the replacement has length identical to the box'es size, as otherwise the operation will fail);
```

File: docs/state.rst [hunk 11]
```diff
@@ -244,3 +246,119 @@ For example:
...
+otherwise, it will create a box containing exactly the same input bytes.
+
+.. code-block:: python
+
+    # create a 42 bytes length box called `poemLine` with content
+    App.box_put(Bytes("poemLine"), Bytes("Of that colossal wreck, boundless and bare"))
+    # write to box `poemLine` with new value
+    App.box_put(Bytes("poemLine"), Bytes("The lone and level sands stretch far away."))
```
Comment (tzaffi) [inline docs/state.rst:289]: very creative!

File: docs/state.rst [hunk 12]
```diff
@@ -244,3 +246,119 @@ For example:
...
+
+Writing to a Box
+~~~~~~~~~~~~~~~~
+
+To write to a box, use :any:`App.box_replace`, or :any:`App.box_put` method.
+
+For :any:`App.box_replace`, it writes bytes of certain length from a start index in a Box.
```
Comment (tzaffi) [inline docs/state.rst:296]:
```suggestion
:any:`App.box_replace` writes bytes of certain length from a start index in a Box.
```

File: docs/state.rst [hunk 13]
```diff
@@ -244,3 +246,119 @@ For example:
...
+The first argument is the box name to write into, the second argument is the starting index to write,
+and the third argument is the replacement bytes. For example:
+
+.. code-block:: python
+
+   # replace 2 bytes to `Ne` starting from 0st bytes in a box with name `wordleBox`
```
Comment (tzaffi) [inline docs/state.rst:302]:
```suggestion
   # replace in the box named `wordleBox` by the 2 bytes `Ne` starting from the 0'th byte
```

File: docs/state.rst [hunk 14]
```diff
@@ -244,3 +246,119 @@ For example:
...
+   App.box_replace(Bytes("wordleBox"), Int(0), Bytes("Ne"))
+
+For :any:`App.box_put`, it writes the full content to a pre-existing box.
```
Comment (tzaffi) [inline docs/state.rst:305]:
```suggestion
:any:`App.box_put` writes the full contents to a pre-existing box.
```

Comment (tzaffi) [reply docs/state.rst:305]: nit: maybe this section is a bit redundant given the example for `box_put` previously provided? But it's also arguable to keep this example given that this section is about "writing". A middle ground would be to just mention that ":any:`App.box_put` can also be used for writing a box'es entire contents, as seen in the example above."

**Review by tzaffi — COMMENTED** [review]
Looks good overall.

Mostly a slew of minor edits.

Also, if you find that it's a value add, please consider including tests for all the new `App` methods inside the file `app_test.py`.

File: pyteal/ast/box_test.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
+from typing import Tuple
+
+import pytest
+import pyteal as pt
+
+teal7Options = pt.CompileOptions(version=7)
+teal8Options = pt.CompileOptions(version=8)
+
+
+def test_compile_version_and_type():
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/box_test.py]: good find, just changed as you suggested

File: pyteal/ast/box_test.py [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+    actual, _ = expr.__teal__(teal8Options)
+    actual.addIncoming()
+    actual = pt.TealBlock.NormalizeBlocks(actual)
+
+    assert expected == actual
+
+
+def test_box_length():
+    name_arg: pt.Expr = pt.Bytes("eineName")
+    expr: pt.MultiValue = pt.BoxLen(name_arg)
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/box_test.py]: ugh the story behind is: `BoxLen` used to be implemented based on `MultiValue`, but the code was not cleaned here 😅 my bad.

File: docs/state.rst [hunk 17]
```diff
@@ -244,3 +246,119 @@ For example:
...
+To create a box, use :any:`App.box_create`, or :any:`App.box_put` method.
+
+For :any:`App.box_create`, the first argument is the box name, and the second argument is the byte size to be allocated.
+
+For example:
+
+.. code-block:: python
+
+    # Allocate a box called "BoxForAlice" of byte size 100
+    App.box_create(Bytes("BoxForAlice"), Int(0xA11CE))
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst:275]: oh boy, hex range gets wild. I will trim the number down

File: docs/state.rst (hunk 8 above)
Comment (ahangsu) [reply docs/state.rst:259]: it just reminds me of my advisor and I are working through paper... he's picking verbal error from each line of my writing... quite a nostalgic feeling, but probably quite painful for my advisor

File: pyteal/ast/maybe.py [hunk 18]
```diff
@@ -27,8 +31,22 @@ def __init__(
             immediate_args (optional): Immediate arguments for the op. Defaults to None.
             args (optional): Stack arguments for the op. Defaults to None.
         """
+
+        def local_version_check(option: "CompileOptions"):
```
[the remark this answers is not in the document]
Comment (tzaffi) [reply pyteal/ast/maybe.py:35]: Was the original intent to do a compile check based on the fact that `MultiValue()` was being employed for these TEAL8 opcodes? Now that `MaybeValue` is being used instead, does it still make sense to have `MultiValue` have this verstion checking, or should it all be done in `MaybeValue` ?

[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/maybe.py:35]: > Was the original intent to do a compile check based on the fact that `MultiValue()` was being employed for these TEAL8 opcodes?

If I am understanding correctly, do you mean the purpose of introducing such compile check is because `MultiValue` was the first implementation of these `box` related `Teal8` opcodes?

If so, then yes. I think `MaybeValue`, which is inherited from `MultiValue`, does not have a compile check, seems flawed.

> Now that `MaybeValue` is being used instead, does it still make sense to have `MultiValue` have this version checking, or should it all be done in `MaybeValue` ?

I believe this it is still making sense, for `MaybeValue` is a 2-return-value special case in general `MultiValue` settings.

I might be understanding it wrongly, but if you are suggesting all the version checking in `MaybeValue` can be applying to all cases in `MultiValue`, I think that will hurt some implementation in `ecdsa`, for they use `verifyFieldVersion` rather than `verifyTealVersion`. Such flexibility seems to be still needed?

File: docs/state.rst (hunk 9 above)
Comment (ahangsu) [reply docs/state.rst]: I just changed the example code block to match with what comments suggested.

File: docs/state.rst (hunk 14 above)
Comment (ahangsu) [reply docs/state.rst]: indeed, it is redundant for `box_put`, as it was previously stated. I was thinking of a way to circumvent re-writing same paragraph cross different sections.

The middle ground suggestion makes sense to me, I will adapt to that suggestion and reflect in the document.

File: docs/state.rst [hunk 19]
```diff
@@ -13,14 +13,16 @@ slices. There are multiple types of state that an application can use.
...
+================== ======================= ======================== ======================== ===================== ======================
+Context            Create                  Write                    Read                     Delete                Check If Exists
+================== ======================= ======================== ======================== ===================== ======================
+Current App Global                         :any:`App.globalPut`     :any:`App.globalGet`     :any:`App.globalDel`  :any:`App.globalGetEx`
+Current App Local                          :any:`App.localPut`      :any:`App.localGet`      :any:`App.localDel`   :any:`App.localGetEx`
+Other App Global                                                    :any:`App.globalGetEx`                         :any:`App.globalGetEx`
+Other App Local                                                     :any:`App.localGetEx`                          :any:`App.localGetEx`
+Current App Boxes  | :any:`App.box_create` | :any:`App.box_put`     | :any:`App.box_extract` :any:`App.box_delete` :any:`App.box_length`
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst:23]: yes, by nature `App.box_get` can be used to check box existence, for it returns 2 values on stack, one of them is box existence.

File: docs/state.rst [hunk 20]
```diff
@@ -244,3 +246,119 @@ For example:
...
+Deleting a Box
+~~~~~~~~~~~~~~
+
+To delete a box, use :any:`App.box_delete` method. The only argument is the box name. For example:
+
+.. code-block:: python
+
+    App.box_delete(Bytes("boxToRemove"))
+
+Checking if a Box Exists
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst]: 👍

File: docs/state.rst [hunk 21]
```diff
@@ -244,3 +246,138 @@ For example:
...
+For example:
+
+.. code-block:: python
+
+   # search for the box length for box `someBox`, and get the bool value for box existence
+   App.box_length(Bytes("someBox")).hasValue()
+
+.. note::
+
+   :any:`App.box_get` can also check the existence of a box, mentioned in `Reading from a Box`_.
```
Comment (tzaffi) [inline docs/state.rst:383]:
```suggestion
   :any:`App.box_get` can also check the existence of a box as mentioned in `Reading from a Box`_.
```

File: pyteal/ast/box_test.py [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+    actual.addIncoming()
+    actual = pt.TealBlock.NormalizeBlocks(actual)
+
+    with pt.TealComponent.Context.ignoreExprEquality():
+        assert expected == actual
+
+
+def test_box_get():
+    name_arg: pt.Expr = pt.Bytes("eineName")
+    expr: pt.MultiValue = pt.BoxGet(name_arg)
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply pyteal/ast/box_test.py]: oh no, I overlooked.

File: docs/state.rst (hunk 21 above)
Comment (ahangsu) [reply docs/state.rst]: 👍

**Review by tzaffi — APPROVED** [review]

File: docs/state.rst [hunk 23]
```diff
@@ -244,3 +246,138 @@ For example:
...
+   It is recommended that *before* app deletion, all box storage be deleted, and funds previously allocated to the MBR be withdrawn.
+
+Box sizes and names cannot be changed after initial allocation, but they can be deleted and re-allocated.
+Boxes are only visible to the application itself; in other words, an application cannot read from or write to another application's boxes on-chain.
+
+Boxes can be manipulated as follows:
+
+.. _Create Boxes:
+
+Create Boxes
```
Comment (michaeldiamant) [inline docs/state.rst:269]: @ahangsu Minor:  For consistency with the sections' headers, I propose a minor tweak.

```suggestion
.. _Creating Boxes:

Creating Boxes
```

File: docs/state.rst [hunk 24]
```diff
@@ -244,3 +246,138 @@ For example:
...
+    App.box_create(Bytes("BoxA"), Int(100))
+    # Allocate a box called "BoxB" of byte size 90
+    App.box_create(Bytes("BoxB"), Int(90)
+
+For :any:`App.box_put`, the first argument is the box name to create or to write to, and the second argument is the bytes to write.
+
+.. note::
+
+   If the box exists, then :any:`App.box_put` will write the contents to the box
+   (assuming the replacement has length **identical** to the box'es byte size, as otherwise the operation will fail);
```
[the remark this answers is not in the document]
Comment (michaeldiamant) [reply docs/state.rst:291]: @ahangsu Alternatively and optionally, the phrasing can be made more concise:

If the box exists, then App.box_put will write the contents to the box (fails when the replacement length is not identical to the box's byte size);

File: docs/state.rst [hunk 25]
```diff
@@ -244,3 +246,138 @@ For example:
...
+
+.. warning::
+
+   If one deletes an application with outstanding boxes, the MBR is not recoverable from the deleted app account.
+   It is recommended that *before* app deletion, all box storage be deleted, and funds previously allocated to the MBR be withdrawn.
+
+Box sizes and names cannot be changed after initial allocation, but they can be deleted and re-allocated.
+Boxes are only visible to the application itself; in other words, an application cannot read from or write to another application's boxes on-chain.
+
+Boxes can be manipulated as follows:
```
Comment (michaeldiamant) [inline docs/state.rst:265]: @ahangsu Minor and optional:  I'd expect an immediate listing following a colon (:) rather than a series of subsections.

I don't feel strongly and you're welcomed to resolve as is.  Alternatives include (a) removing the line or (b) rephrasing (_suggestion_ offers one phrasing).

```suggestion
The following sections explain how to work with boxes.
```

File: docs/state.rst [hunk 26]
```diff
@@ -244,3 +246,138 @@ For example:
         otherAppOtherAccountRole,
         If(otherAppOtherAccountRole.hasValue(), otherAppOtherAccountRole.value(), Bytes("none"))
     ])
+
+Box Storage
+-----------
+
+Box storage consists of key-value pairs that are stored in an application's local context.
+
+The app account's minimum balance requirement (MBR) is increased with each additional box, and each additional byte in the box'es name and allocated size.
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst:255]: 👍

File: docs/state.rst (hunk 23 above)
Comment (ahangsu) [reply docs/state.rst:269]: yes, I will take care of the section replacing.

File: docs/state.rst (hunk 25 above)
Comment (ahangsu) [reply docs/state.rst:265]: I would like to rephrase this as you suggested, feels making more sense.

File: docs/state.rst (hunk 24 above)
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst:291]: I go with the more concise version of rephrasing.

File: docs/state.rst [hunk 27]
```diff
@@ -13,14 +13,16 @@ slices. There are multiple types of state that an application can use.
 State Operation Table
```
[the remark this answers is not in the document]
Comment (ahangsu) [reply docs/state.rst:21]: just insert a table to explain allowed types. let me know if it looks right.

[the remark this answers is not in the document]
Comment (michaeldiamant) [reply docs/state.rst:21]: @ahangsu Resolving - Table looks good.  Thanks for bringing in the feedback.

**Review by michaeldiamant — APPROVED** [review]
@ahangsu Thanks for your efforts to bring Box opcodes to PyTeal! ☕
